File: Algorithm/Python/257.binary-tree-paths.py
#
# @lc app=leetcode id=257 lang=python
#
# [257] Binary Tree Paths
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def binaryTreePaths(self, root):
        """
        :type root: TreeNode
        :rtype: List[str]
        """
        result, tmp = [], []
        if root == None:
            return result
        self.dfs(result, tmp, root)
        return result

    def dfs(self, result, tmp, node):
        tmp.append(str(node.val))
        if node.left == None and node.right == None:
            result.append("->".join(tmp))
        else:
            for _node in [node.left, node.right]:
                if _node != None:
                    self.dfs(result, tmp, _node)
        tmp.pop()
# @lc code=end

File: Algorithm/C++/257.binary-tree-paths.cpp
/*
 * @lc app=leetcode id=257 lang=cpp
 *
 * [257] Binary Tree Paths
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
#include <vector>
#include <string>

class Solution {
public:
    std::vector<std::string> binaryTreePaths(TreeNode* root) {
        std::vector<std::string> result;
        std::string tmp;
        if (root == nullptr) {
            return result;
        }
        dfs(result, tmp, root);
        return result;
    }

private:
    void dfs(std::vector<std::string>& result, std::string tmp, TreeNode* node) {
        if (tmp.size() > 0) {
            tmp += "->";
        }
        tmp += std::to_string(node->val);
        if (node->left == nullptr && node->right == nullptr) {
            result.push_back(tmp);
        } else {
            if (node->left != nullptr) {
                dfs(result, tmp, node->left);
            }
            if (node->right != nullptr) {
                dfs(result, tmp, node->right);
            }
        }
    }
};
// @lc code=end

